{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d76043a2",
   "metadata": {},
   "source": [
    "# Node prioritisation in Leukemia with regulatory and metabolic networks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "971d31b6",
   "metadata": {},
   "source": [
    "In this tutorial, we illustrate how MultiXrank can be used to prioritise genes and drugs of interest in leukemia. \n",
    "We build upon a multilayer network composed of:\n",
    "    \n",
    "- a gene multiplex network, encoding proteins associations in protein-protrin interactions, protein complexes and pathways; \n",
    "- a drug multiplex network, encoding clinically reported drug-drug interactions, experimental and predicted drug combinations, and phramacological interactions.\n",
    "\n",
    "In this tutorial, we complete this multilayer network with two additional types of interactions:\n",
    "\t\n",
    "- **Network of metabolic reactions**: We dowloaded the human genome-scale metabolic model from the Human-GEM database and create a network of metabolic reactions.  Specifically, this additional layer contains nodes representing specific human metabolic reactions. Directed edges connect two reactions when the product of one reaction corresponds to the reactants of another. We connect the metabolic reaction network to the gene multiplex through bipartite reactions for each gene known to participate in a metabolic reaction.\t\n",
    "- **Regulatory network of Transcription Factors and their targets**: We add to the existing gene multiplex a new layer featuring directed interactions representing Transcription Factors (TF) and their regulated targets in leukemia. This network was constructed using data from the GRNdb database.\n",
    "\n",
    "Finally, we run MultiXrank on this novel multilayer netork to prioritise genes and drugs of interest in leukemia, using the HRAS gene and the Tipifarnib drug as seeds."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18b82521",
   "metadata": {},
   "source": [
    "# 1) Network of metabolic reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79ab6e1f",
   "metadata": {},
   "source": [
    "The human-GEM was downaloaded from: https://github.com/SysBioChalmers/Human-GEM\n",
    "\n",
    "    J. L. Robinson, P. Kocabas, H. Wang, P.-E. Cholley, et al. An atlas of human metabolism. Sci. Signal. 13, eaaz1482 (2020). doi:10.1126/scisignal.aaz1482\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "067a8fe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import networkx as nx\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "78ff35a5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Rxn name</th>\n",
       "      <th>Formula</th>\n",
       "      <th>Gene-reaction association</th>\n",
       "      <th>LB</th>\n",
       "      <th>UB</th>\n",
       "      <th>Objective</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAR03905</td>\n",
       "      <td>MAM01796c + MAM02552c  -&gt; MAM01249c + MAM02039...</td>\n",
       "      <td>ENSG00000147576 or ENSG00000172955 or ENSG0000...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MAR03907</td>\n",
       "      <td>MAM01796c + MAM02554c  -&gt; MAM01249c + MAM02039...</td>\n",
       "      <td>ENSG00000117448</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAR04097</td>\n",
       "      <td>MAM01252c + MAM01371c + MAM01597c  -&gt; MAM01261...</td>\n",
       "      <td>ENSG00000131069</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAR04099</td>\n",
       "      <td>MAM01252m + MAM01371m + MAM01597m  -&gt; MAM01261...</td>\n",
       "      <td>ENSG00000111058 or ENSG00000154930</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAR04108</td>\n",
       "      <td>MAM01257c + MAM01597c  -&gt; MAM01261c + MAM01334...</td>\n",
       "      <td>ENSG00000131069</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12990</th>\n",
       "      <td>MAR20168</td>\n",
       "      <td>MAM00209c  -&gt; MAM02393m</td>\n",
       "      <td>ENSG00000144182</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12991</th>\n",
       "      <td>MAR20169</td>\n",
       "      <td>MAM01412c + MAM02040c  -&gt; MAM01410c + MAM01597...</td>\n",
       "      <td>ENSG00000172497</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12992</th>\n",
       "      <td>MAR20170</td>\n",
       "      <td>MAM02040c + MAM02644c  -&gt; MAM01597c + MAM02039...</td>\n",
       "      <td>ENSG00000097021</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12993</th>\n",
       "      <td>MAR20171</td>\n",
       "      <td>MAM01650c + MAM02040c  -&gt; MAM01597c + MAM01648...</td>\n",
       "      <td>ENSG00000097021</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12994</th>\n",
       "      <td>MAR20172</td>\n",
       "      <td>MAM02040c + MAM03116c  -&gt; MAM01597c + MAM02039...</td>\n",
       "      <td>ENSG00000097021</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12995 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Rxn name                                            Formula  \\\n",
       "0      MAR03905  MAM01796c + MAM02552c  -> MAM01249c + MAM02039...   \n",
       "1      MAR03907  MAM01796c + MAM02554c  -> MAM01249c + MAM02039...   \n",
       "2      MAR04097  MAM01252c + MAM01371c + MAM01597c  -> MAM01261...   \n",
       "3      MAR04099  MAM01252m + MAM01371m + MAM01597m  -> MAM01261...   \n",
       "4      MAR04108  MAM01257c + MAM01597c  -> MAM01261c + MAM01334...   \n",
       "...         ...                                                ...   \n",
       "12990  MAR20168                           MAM00209c  -> MAM02393m    \n",
       "12991  MAR20169  MAM01412c + MAM02040c  -> MAM01410c + MAM01597...   \n",
       "12992  MAR20170  MAM02040c + MAM02644c  -> MAM01597c + MAM02039...   \n",
       "12993  MAR20171  MAM01650c + MAM02040c  -> MAM01597c + MAM01648...   \n",
       "12994  MAR20172  MAM02040c + MAM03116c  -> MAM01597c + MAM02039...   \n",
       "\n",
       "                               Gene-reaction association   LB      UB  \\\n",
       "0      ENSG00000147576 or ENSG00000172955 or ENSG0000...  0.0  1000.0   \n",
       "1                                        ENSG00000117448  0.0  1000.0   \n",
       "2                                        ENSG00000131069  0.0  1000.0   \n",
       "3                     ENSG00000111058 or ENSG00000154930  0.0  1000.0   \n",
       "4                                        ENSG00000131069  0.0  1000.0   \n",
       "...                                                  ...  ...     ...   \n",
       "12990                                    ENSG00000144182  0.0  1000.0   \n",
       "12991                                    ENSG00000172497  0.0  1000.0   \n",
       "12992                                    ENSG00000097021  0.0  1000.0   \n",
       "12993                                    ENSG00000097021  0.0  1000.0   \n",
       "12994                                    ENSG00000097021  0.0  1000.0   \n",
       "\n",
       "       Objective  \n",
       "0            0.0  \n",
       "1            0.0  \n",
       "2            0.0  \n",
       "3            0.0  \n",
       "4            0.0  \n",
       "...          ...  \n",
       "12990        0.0  \n",
       "12991        0.0  \n",
       "12992        0.0  \n",
       "12993        0.0  \n",
       "12994        0.0  \n",
       "\n",
       "[12995 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"Human-GEM.txt\", sep=\"\\t\")\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61b555ad",
   "metadata": {},
   "source": [
    "## Create a dictionary of reactions reactants and products"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "db0462aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "separators = re.compile(r'\\s*(->|<=>)\\s*')\n",
    "\n",
    "mar_reactants_dict = {}\n",
    "mar_products_dict = {}\n",
    "\n",
    "for index, row in data.iterrows():\n",
    "    rxn_id = row['Rxn name']\n",
    "    \n",
    "    reactants, sep, products = re.split(separators, row['Formula'])\n",
    "    \n",
    "    mar_reactants_dict[rxn_id] = reactants.split(' + ')\n",
    "    \n",
    "    mar_products_dict[rxn_id] = products.split(' + ')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a3636f3",
   "metadata": {},
   "source": [
    "## Create the network of metabolic reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad5270de",
   "metadata": {},
   "source": [
    "Directed edge list of reactions: a directed edge is created between two reactions when the product of the first reaction is used in the reactants of the second reaction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f770e533",
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_list = []\n",
    "\n",
    "# Iterate through reaction IDs\n",
    "for reaction_id2, products2 in mar_products_dict.items():\n",
    "    for reaction_id1, reactants1 in mar_reactants_dict.items():\n",
    "        common_metabolites = set(reactants1) & set(products2)\n",
    "        if common_metabolites:\n",
    "            edge_list.append((reaction_id2, reaction_id1, list(common_metabolites)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d94ae78f",
   "metadata": {},
   "outputs": [],
   "source": [
    "edges = pd.DataFrame(edge_list, columns=['Reaction1', 'Reaction2', 'CommonMetabolites'])\n",
    "\n",
    "edges[['Reaction1', 'Reaction2']].to_csv('multiplex/3/mar_mar.tsv', sep='\\t', index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1feba468",
   "metadata": {},
   "source": [
    "## Create a dictionary of gene-reaction associations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f345c146",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mar_gene_dict = {}\n",
    "for index, row in data.iterrows():\n",
    "    rxn_id = row['Rxn name']\n",
    "    \n",
    "    if isinstance(row['Gene-reaction association'], str):\n",
    "        # Extract all strings starting with ENSG and ending with a space\n",
    "        genes = re.findall(r'\\bENSG[0-9]+\\b', row['Gene-reaction association'])\n",
    "        \n",
    "        mar_gene_dict[rxn_id] = set(genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55bfacd6",
   "metadata": {},
   "source": [
    "Convert ensembl ids to gene symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "97139702",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>geneENSTID</th>\n",
       "      <th>geneENSPID</th>\n",
       "      <th>geneUniProtID</th>\n",
       "      <th>geneSymbols</th>\n",
       "      <th>geneEntrezID</th>\n",
       "      <th>geneNames</th>\n",
       "      <th>geneAliases</th>\n",
       "      <th>compartments</th>\n",
       "      <th>compDataSource</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENST00000466152;ENST00000371582;ENST0000068304...</td>\n",
       "      <td>ENSP00000507119;ENSP00000360638;ENSP0000050698...</td>\n",
       "      <td>O60762</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>8813</td>\n",
       "      <td>dolichyl-phosphate mannosyltransferase subunit...</td>\n",
       "      <td>CDGIE;MPDS</td>\n",
       "      <td>Endoplasmic reticulum</td>\n",
       "      <td>SwissProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000001036</td>\n",
       "      <td>ENST00000002165;ENST00000451668;ENST00000367585</td>\n",
       "      <td>ENSP00000002165;ENSP00000398119</td>\n",
       "      <td>Q9BTY2</td>\n",
       "      <td>FUCA2</td>\n",
       "      <td>2519</td>\n",
       "      <td>alpha-L-fucosidase 2</td>\n",
       "      <td>dJ20N2.5;MGC1314</td>\n",
       "      <td>Extracellular</td>\n",
       "      <td>DeepLoc2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000001084</td>\n",
       "      <td>ENST00000509541;ENST00000650454;ENST0000061692...</td>\n",
       "      <td>ENSP00000495056;ENSP00000497574;ENSP0000048275...</td>\n",
       "      <td>P48506</td>\n",
       "      <td>GCLC</td>\n",
       "      <td>2729</td>\n",
       "      <td>glutamate-cysteine ligase catalytic subunit</td>\n",
       "      <td>GCS;GLCL;GLCLC</td>\n",
       "      <td>Cytosol;Nucleus</td>\n",
       "      <td>CellAtlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000001630</td>\n",
       "      <td>ENST00000003100;ENST00000450723;ENST0000042286...</td>\n",
       "      <td>ENSP00000003100;ENSP00000406757;ENSP00000394268</td>\n",
       "      <td>Q16850</td>\n",
       "      <td>CYP51A1</td>\n",
       "      <td>1595</td>\n",
       "      <td>cytochrome P450 family 51 subfamily A member 1</td>\n",
       "      <td>CP51;CYP51;CYPL1;LDM;P450-14DM;P450L1</td>\n",
       "      <td>Endoplasmic reticulum</td>\n",
       "      <td>SwissProt;CellAtlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000002549</td>\n",
       "      <td>ENST00000226299;ENST00000618908;ENST0000050849...</td>\n",
       "      <td>ENSP00000226299;ENSP00000481000;ENSP0000047602...</td>\n",
       "      <td>P28838</td>\n",
       "      <td>LAP3</td>\n",
       "      <td>51056</td>\n",
       "      <td>leucine aminopeptidase 3</td>\n",
       "      <td>LAP;LAPEP;PEPS</td>\n",
       "      <td>Nucleus;Cytosol</td>\n",
       "      <td>SwissProt;CellAtlas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2884</th>\n",
       "      <td>ENSG00000114120</td>\n",
       "      <td>ENST00000324194;ENST00000446041;ENST0000050742...</td>\n",
       "      <td>ENSP00000320688;ENSP00000401938;ENSP0000042147...</td>\n",
       "      <td>Q96CQ1</td>\n",
       "      <td>SLC25A36</td>\n",
       "      <td>55186</td>\n",
       "      <td>solute carrier family 25 member 36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2885</th>\n",
       "      <td>ENSG00000159445</td>\n",
       "      <td>ENST00000368814.8;ENST00000489410.1;ENST000004...</td>\n",
       "      <td>ENSP00000357804.3;ENSP00000433304.1;ENSP000004...</td>\n",
       "      <td>Q5T1C6</td>\n",
       "      <td>THEM4</td>\n",
       "      <td>117145</td>\n",
       "      <td>thioesterase superfamily member 4</td>\n",
       "      <td>CTMP</td>\n",
       "      <td>Cytosol;Inner mitochondria</td>\n",
       "      <td>SwissProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2886</th>\n",
       "      <td>ENSG00000196407</td>\n",
       "      <td>ENST00000368817.10;ENST00000453881.2</td>\n",
       "      <td>ENSP00000357807.4;ENSP00000406809.2</td>\n",
       "      <td>Q8N1Q8</td>\n",
       "      <td>THEM5</td>\n",
       "      <td>284486</td>\n",
       "      <td>thioesterase superfamily member 5</td>\n",
       "      <td>ACOT15</td>\n",
       "      <td>Inner mitochondria</td>\n",
       "      <td>SwissProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2887</th>\n",
       "      <td>ENSG00000013306</td>\n",
       "      <td>ENST00000377095;ENST00000225308;ENST0000059019...</td>\n",
       "      <td>ENSP00000366299;ENSP00000225308;ENSP0000046768...</td>\n",
       "      <td>Q9BZJ4</td>\n",
       "      <td>SLC25A39</td>\n",
       "      <td>51629</td>\n",
       "      <td>solute carrier family 25 member 39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2888</th>\n",
       "      <td>ENSG00000075303</td>\n",
       "      <td>ENST00000341119;ENST00000429674;ENST0000044436...</td>\n",
       "      <td>ENSP00000344831;ENSP00000401473;ENSP0000040556...</td>\n",
       "      <td>Q8TBP6</td>\n",
       "      <td>SLC25A40</td>\n",
       "      <td>55972</td>\n",
       "      <td>solute carrier family 25 member 40</td>\n",
       "      <td>MCFP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2889 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                genes                                         geneENSTID  \\\n",
       "0     ENSG00000000419  ENST00000466152;ENST00000371582;ENST0000068304...   \n",
       "1     ENSG00000001036    ENST00000002165;ENST00000451668;ENST00000367585   \n",
       "2     ENSG00000001084  ENST00000509541;ENST00000650454;ENST0000061692...   \n",
       "3     ENSG00000001630  ENST00000003100;ENST00000450723;ENST0000042286...   \n",
       "4     ENSG00000002549  ENST00000226299;ENST00000618908;ENST0000050849...   \n",
       "...               ...                                                ...   \n",
       "2884  ENSG00000114120  ENST00000324194;ENST00000446041;ENST0000050742...   \n",
       "2885  ENSG00000159445  ENST00000368814.8;ENST00000489410.1;ENST000004...   \n",
       "2886  ENSG00000196407               ENST00000368817.10;ENST00000453881.2   \n",
       "2887  ENSG00000013306  ENST00000377095;ENST00000225308;ENST0000059019...   \n",
       "2888  ENSG00000075303  ENST00000341119;ENST00000429674;ENST0000044436...   \n",
       "\n",
       "                                             geneENSPID geneUniProtID  \\\n",
       "0     ENSP00000507119;ENSP00000360638;ENSP0000050698...        O60762   \n",
       "1                       ENSP00000002165;ENSP00000398119        Q9BTY2   \n",
       "2     ENSP00000495056;ENSP00000497574;ENSP0000048275...        P48506   \n",
       "3       ENSP00000003100;ENSP00000406757;ENSP00000394268        Q16850   \n",
       "4     ENSP00000226299;ENSP00000481000;ENSP0000047602...        P28838   \n",
       "...                                                 ...           ...   \n",
       "2884  ENSP00000320688;ENSP00000401938;ENSP0000042147...        Q96CQ1   \n",
       "2885  ENSP00000357804.3;ENSP00000433304.1;ENSP000004...        Q5T1C6   \n",
       "2886                ENSP00000357807.4;ENSP00000406809.2        Q8N1Q8   \n",
       "2887  ENSP00000366299;ENSP00000225308;ENSP0000046768...        Q9BZJ4   \n",
       "2888  ENSP00000344831;ENSP00000401473;ENSP0000040556...        Q8TBP6   \n",
       "\n",
       "     geneSymbols geneEntrezID  \\\n",
       "0           DPM1         8813   \n",
       "1          FUCA2         2519   \n",
       "2           GCLC         2729   \n",
       "3        CYP51A1         1595   \n",
       "4           LAP3        51056   \n",
       "...          ...          ...   \n",
       "2884    SLC25A36        55186   \n",
       "2885       THEM4       117145   \n",
       "2886       THEM5       284486   \n",
       "2887    SLC25A39        51629   \n",
       "2888    SLC25A40        55972   \n",
       "\n",
       "                                              geneNames  \\\n",
       "0     dolichyl-phosphate mannosyltransferase subunit...   \n",
       "1                                  alpha-L-fucosidase 2   \n",
       "2           glutamate-cysteine ligase catalytic subunit   \n",
       "3        cytochrome P450 family 51 subfamily A member 1   \n",
       "4                              leucine aminopeptidase 3   \n",
       "...                                                 ...   \n",
       "2884                 solute carrier family 25 member 36   \n",
       "2885                  thioesterase superfamily member 4   \n",
       "2886                  thioesterase superfamily member 5   \n",
       "2887                 solute carrier family 25 member 39   \n",
       "2888                 solute carrier family 25 member 40   \n",
       "\n",
       "                                geneAliases                compartments  \\\n",
       "0                                CDGIE;MPDS       Endoplasmic reticulum   \n",
       "1                          dJ20N2.5;MGC1314               Extracellular   \n",
       "2                            GCS;GLCL;GLCLC             Cytosol;Nucleus   \n",
       "3     CP51;CYP51;CYPL1;LDM;P450-14DM;P450L1       Endoplasmic reticulum   \n",
       "4                            LAP;LAPEP;PEPS             Nucleus;Cytosol   \n",
       "...                                     ...                         ...   \n",
       "2884                                    NaN                         NaN   \n",
       "2885                                   CTMP  Cytosol;Inner mitochondria   \n",
       "2886                                 ACOT15          Inner mitochondria   \n",
       "2887                                    NaN                         NaN   \n",
       "2888                                   MCFP                         NaN   \n",
       "\n",
       "           compDataSource  \n",
       "0               SwissProt  \n",
       "1                DeepLoc2  \n",
       "2               CellAtlas  \n",
       "3     SwissProt;CellAtlas  \n",
       "4     SwissProt;CellAtlas  \n",
       "...                   ...  \n",
       "2884                  NaN  \n",
       "2885            SwissProt  \n",
       "2886            SwissProt  \n",
       "2887                  NaN  \n",
       "2888                  NaN  \n",
       "\n",
       "[2889 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_csv(\"genes.tsv\", sep=\"\\t\") # Downloaded from Human-GEM\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "652758b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensg_to_symbol = dict(zip(genes['genes'], genes['geneSymbols']))\n",
    "\n",
    "mapped_mar_gene_dict = {mar_id: [ensg_to_symbol.get(ensg, ensg) for ensg in engrn] for mar_id, engrn in mar_gene_dict.items()}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "db3576a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MAR_ID</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAR03905</td>\n",
       "      <td>ADH6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MAR03905</td>\n",
       "      <td>ADH1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAR03905</td>\n",
       "      <td>ADH7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAR03905</td>\n",
       "      <td>ADH1C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAR03905</td>\n",
       "      <td>ADH5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25889</th>\n",
       "      <td>MAR20168</td>\n",
       "      <td>LIPT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25890</th>\n",
       "      <td>MAR20169</td>\n",
       "      <td>ACOT12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25891</th>\n",
       "      <td>MAR20170</td>\n",
       "      <td>ACOT7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25892</th>\n",
       "      <td>MAR20171</td>\n",
       "      <td>ACOT7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25893</th>\n",
       "      <td>MAR20172</td>\n",
       "      <td>ACOT7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25894 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         MAR_ID    Gene\n",
       "0      MAR03905    ADH6\n",
       "1      MAR03905   ADH1A\n",
       "2      MAR03905    ADH7\n",
       "3      MAR03905   ADH1C\n",
       "4      MAR03905    ADH5\n",
       "...         ...     ...\n",
       "25889  MAR20168   LIPT1\n",
       "25890  MAR20169  ACOT12\n",
       "25891  MAR20170   ACOT7\n",
       "25892  MAR20171   ACOT7\n",
       "25893  MAR20172   ACOT7\n",
       "\n",
       "[25894 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bipartites = [(mar_id, gene) for mar_id, genes in mapped_mar_gene_dict.items() for gene in genes]\n",
    "\n",
    "bipartites = pd.DataFrame(bipartites, columns=['MAR_ID', 'Gene'])\n",
    "bipartites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "628dd2fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "bipartites.to_csv('bipartite/3_1.tsv', sep='\\t', index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5731461b",
   "metadata": {},
   "source": [
    "# 2) Network of Transcription Factor - target interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5c6c24c",
   "metadata": {},
   "source": [
    "The data was downloaded from http://www.grndb.com/\n",
    "\n",
    "    Fang et al. GRNdb: decoding the gene regulatory networks in diverse human and mouse conditions, Nucleic Acids Research, 2020 (DOI: 10.1093/nar/gkaa995)\n",
    "\n",
    "    Li et al. Single-cell transcriptomic analysis reveals dynamic alternative splicing and gene regulatory networks among pancreatic islets, Science China Life Sciences, 2020 (DOI: 10.1007/s11427-020-1711-x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "225951a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>gene</th>\n",
       "      <th>bestMotif</th>\n",
       "      <th>NES</th>\n",
       "      <th>Genie3Weight</th>\n",
       "      <th>Confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>GNAT1</td>\n",
       "      <td>hdpi__ACF</td>\n",
       "      <td>3.45</td>\n",
       "      <td>0.007303</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>RPL23AP87</td>\n",
       "      <td>hdpi__ACF</td>\n",
       "      <td>3.45</td>\n",
       "      <td>0.008207</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AHCTF1</td>\n",
       "      <td>AHCTF1</td>\n",
       "      <td>cisbp__M0131</td>\n",
       "      <td>3.03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AHCTF1</td>\n",
       "      <td>AP1G1</td>\n",
       "      <td>cisbp__M0131</td>\n",
       "      <td>3.03</td>\n",
       "      <td>0.009322</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AHCTF1</td>\n",
       "      <td>DYRK1A</td>\n",
       "      <td>cisbp__M0131</td>\n",
       "      <td>3.03</td>\n",
       "      <td>0.014870</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115375</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>TECPR2</td>\n",
       "      <td>cisbp__M0507</td>\n",
       "      <td>3.09</td>\n",
       "      <td>0.011065</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115376</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>TMEM110</td>\n",
       "      <td>cisbp__M0507</td>\n",
       "      <td>3.09</td>\n",
       "      <td>0.008685</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115377</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>TRAK1</td>\n",
       "      <td>cisbp__M0507</td>\n",
       "      <td>3.09</td>\n",
       "      <td>0.008408</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115378</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>TRIM72</td>\n",
       "      <td>cisbp__M0507</td>\n",
       "      <td>3.09</td>\n",
       "      <td>0.003858</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115379</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ZBTB43</td>\n",
       "      <td>cisbp__M0507</td>\n",
       "      <td>3.09</td>\n",
       "      <td>0.006579</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>115380 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            TF       gene     bestMotif   NES  Genie3Weight Confidence\n",
       "0         A1CF      GNAT1     hdpi__ACF  3.45      0.007303       High\n",
       "1         A1CF  RPL23AP87     hdpi__ACF  3.45      0.008207       High\n",
       "2       AHCTF1     AHCTF1  cisbp__M0131  3.03           NaN       High\n",
       "3       AHCTF1      AP1G1  cisbp__M0131  3.03      0.009322       High\n",
       "4       AHCTF1     DYRK1A  cisbp__M0131  3.03      0.014870       High\n",
       "...        ...        ...           ...   ...           ...        ...\n",
       "115375    ZXDC     TECPR2  cisbp__M0507  3.09      0.011065       High\n",
       "115376    ZXDC    TMEM110  cisbp__M0507  3.09      0.008685       High\n",
       "115377    ZXDC      TRAK1  cisbp__M0507  3.09      0.008408       High\n",
       "115378    ZXDC     TRIM72  cisbp__M0507  3.09      0.003858       High\n",
       "115379    ZXDC     ZBTB43  cisbp__M0507  3.09      0.006579       High\n",
       "\n",
       "[115380 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf = pd.read_csv('AML_TCGA-regulons.txt', sep=\"\\t\")\n",
    "tf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "955c6644",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf[[\"TF\", \"gene\"]].to_csv(\"multiplex/1/tf_target.tsv\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bb21bf8",
   "metadata": {},
   "source": [
    "# 3)  Running MultiXrank"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ac5ccff",
   "metadata": {},
   "source": [
    "### 3.1) Creating the seed file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd930657",
   "metadata": {},
   "source": [
    "To indicate the set of seeds to be used in the Random Walk with Restart, we create a ```seeds.txt``` file containing the ids of our two nodes of interest:\n",
    "```\n",
    "    HRAS\n",
    "    DB04960\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b67285f",
   "metadata": {},
   "source": [
    "### 3.2) Creating the configuration file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "200b4e6c",
   "metadata": {},
   "source": [
    "We also have to create the configuration file, ```config_full.yml``` that holds the parameters to be used when running MultiXrank. The documentation to create this configuration file is available here: https://multixrank-doc.readthedocs.io/en/latest/content/reference.html#configuration-and-network-files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7efc66e",
   "metadata": {},
   "source": [
    "```\n",
    "seed: seeds.txt\n",
    "r: 0.7 # Restart probability\n",
    "# self_loops: Are self-loops allowed? 1 yes, 0 not\n",
    "self_loops: 0\n",
    "# eta: Restart probability for each multiplex. Must sum up to one.\n",
    "eta: [1/3,1/3, 1/3]\n",
    "#4. Inter multiplex Networks Jump Probability (lambda)\n",
    "lamb:\n",
    "    - [1/3, 1/3, 1/3]\n",
    "    - [1/3, 1/3, 1/3]\n",
    "    - [1/3, 1/3, 1/3]\n",
    "multiplex:\n",
    "    1:\n",
    "        layers:\n",
    "            - multiplex/1/Complexes.tsv\n",
    "            - multiplex/1/PPI.tsv\n",
    "            - multiplex/1/Reactome.tsv\n",
    "            - multiplex/1/tf_target.tsv # This layer is directed (graph_type = 10)\n",
    "        delta: 0.5\n",
    "        # Graph type: undirected/directed, unweighted/weighted\n",
    "        graph_type: [00, 00, 00, 10]\n",
    "        # Restart probability. Layers Restart Probability (tau)\n",
    "        tau: [1/4, 1/4, 1/4, 1/4]\n",
    "    2:\n",
    "        layers:\n",
    "            - multiplex/2/clinical_drug_interactions.tsv\n",
    "            - multiplex/2/drugbank-chem-chem.tsv\n",
    "            - multiplex/2/experimental_drug_combinations.tsv\n",
    "            - multiplex/2/predicted_drug_combinations.tsv\n",
    "        delta: 0.5\n",
    "        graph_type: [00, 00, 00, 00]\n",
    "        tau: [1/4, 1/4, 1/4, 1/4]\n",
    "    3:\n",
    "        layers:\n",
    "            - multiplex/3/mar_mar.tsv # This layer is directed (graph_type = 10)\n",
    "        delta: 0.5\n",
    "        graph_type: [10]\n",
    "        tau: [1]\n",
    "bipartite:\n",
    "    bipartite/2_1.tsv: {source: 2, 'target': 1, graph_type: 00}\n",
    "    bipartite/3_1.tsv: {source: 3, 'target': 1, graph_type: 00} # The novel bipartite network connecting proteins and metabolic reactions\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e1936e2",
   "metadata": {},
   "source": [
    "### 3.3) Running MultiXrank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bb801823",
   "metadata": {},
   "outputs": [],
   "source": [
    "import multixrank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ed83436d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create MultiXrank object, using the configuration file\n",
    "multixrank_obj = multixrank.Multixrank(config=\"config_full.yml\", wdir='./')\n",
    "\n",
    "# Perform the Random Walk with Restart\n",
    "ranking_df = multixrank_obj.random_walk_rank()\n",
    "\n",
    "# Save the results\n",
    "multixrank_obj.write_ranking(ranking_df, path=\"multiXrank_results/\", top=None)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c596679",
   "metadata": {},
   "source": [
    "# 4)  Exploring MultiXrank results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79e9ca67",
   "metadata": {},
   "source": [
    "### 4.1) Prioritised genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39ac2ad0",
   "metadata": {},
   "source": [
    "The results of the node prioritisation for the gene multiplex are output in file `multiXrank_results/multiplex_1.tsv`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1fe4a44c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>multiplex</th>\n",
       "      <th>node</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HRAS</td>\n",
       "      <td>0.091300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CYP3A4</td>\n",
       "      <td>0.005630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>FNTB</td>\n",
       "      <td>0.005460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ZBTB17</td>\n",
       "      <td>0.001017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>RASGRP1</td>\n",
       "      <td>0.000431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>RAF1</td>\n",
       "      <td>0.000310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>GABPA</td>\n",
       "      <td>0.000297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>RIN1</td>\n",
       "      <td>0.000285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>DGKZ</td>\n",
       "      <td>0.000261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>AURKA</td>\n",
       "      <td>0.000088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>0.000080</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    multiplex     node     score\n",
       "0           1     HRAS  0.091300\n",
       "1           1   CYP3A4  0.005630\n",
       "2           1     FNTB  0.005460\n",
       "3           1   ZBTB17  0.001017\n",
       "4           1  RASGRP1  0.000431\n",
       "5           1     RAF1  0.000310\n",
       "6           1    GABPA  0.000297\n",
       "7           1     RIN1  0.000285\n",
       "8           1     DGKZ  0.000261\n",
       "9           1    AURKA  0.000088\n",
       "10          1     BRAF  0.000080"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_prio = pd.read_csv('multiXrank_results/multiplex_1.tsv', sep=\"\\t\")\n",
    "gene_prio.head(11)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bfb9aa8",
   "metadata": {},
   "source": [
    "### 4.2) Prioritised drugs "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "569d0bfd",
   "metadata": {},
   "source": [
    "The results of the node prioritisation for the gene multiplex are output in file `multiXrank_results/multiplex_2.tsv`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f2247024",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>multiplex</th>\n",
       "      <th>node</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>DB04960</td>\n",
       "      <td>0.104663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>DB00399</td>\n",
       "      <td>0.000335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>DB00773</td>\n",
       "      <td>0.000330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>DB00398</td>\n",
       "      <td>0.000225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>DB00637</td>\n",
       "      <td>0.000166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>DB01380</td>\n",
       "      <td>0.000164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>DB00630</td>\n",
       "      <td>0.000150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>DB01254</td>\n",
       "      <td>0.000086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>DB01268</td>\n",
       "      <td>0.000084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>DB06589</td>\n",
       "      <td>0.000053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2</td>\n",
       "      <td>DB00530</td>\n",
       "      <td>0.000052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    multiplex     node     score\n",
       "0           2  DB04960  0.104663\n",
       "1           2  DB00399  0.000335\n",
       "2           2  DB00773  0.000330\n",
       "3           2  DB00398  0.000225\n",
       "4           2  DB00637  0.000166\n",
       "5           2  DB01380  0.000164\n",
       "6           2  DB00630  0.000150\n",
       "7           2  DB01254  0.000086\n",
       "8           2  DB01268  0.000084\n",
       "9           2  DB06589  0.000053\n",
       "10          2  DB00530  0.000052"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_prio = pd.read_csv('multiXrank_results/multiplex_2.tsv', sep=\"\\t\")\n",
    "drug_prio.head(11)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc58ae8b",
   "metadata": {},
   "source": [
    "### 4.3) Prioritised metabolic reactions "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "600ef6eb",
   "metadata": {},
   "source": [
    "Because we have added a novel layer for metabolic reactions, we are now able to prioritise metabolic reactions associated with our seeds. The results of the node prioritisation for the metabolic reaction monoplex are output in file `multiXrank_results/multiplex_3.tsv`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "921b1312",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>multiplex</th>\n",
       "      <th>node</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>MAR07166</td>\n",
       "      <td>0.002210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>MAR00604</td>\n",
       "      <td>0.000236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>MAR00141</td>\n",
       "      <td>0.000235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>MAR01460</td>\n",
       "      <td>0.000152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>MAR09499</td>\n",
       "      <td>0.000148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3</td>\n",
       "      <td>MAR06555</td>\n",
       "      <td>0.000148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3</td>\n",
       "      <td>MAR07165</td>\n",
       "      <td>0.000147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3</td>\n",
       "      <td>MAR09491</td>\n",
       "      <td>0.000107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3</td>\n",
       "      <td>MAR08816</td>\n",
       "      <td>0.000086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>MAR00158</td>\n",
       "      <td>0.000042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3</td>\n",
       "      <td>MAR01451</td>\n",
       "      <td>0.000019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    multiplex      node     score\n",
       "0           3  MAR07166  0.002210\n",
       "1           3  MAR00604  0.000236\n",
       "2           3  MAR00141  0.000235\n",
       "3           3  MAR01460  0.000152\n",
       "4           3  MAR09499  0.000148\n",
       "5           3  MAR06555  0.000148\n",
       "6           3  MAR07165  0.000147\n",
       "7           3  MAR09491  0.000107\n",
       "8           3  MAR08816  0.000086\n",
       "9           3  MAR00158  0.000042\n",
       "10          3  MAR01451  0.000019"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reac_prio = pd.read_csv('multiXrank_results/multiplex_3.tsv', sep=\"\\t\")\n",
    "reac_prio.head(11)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e65c60a1",
   "metadata": {},
   "source": [
    "### 4.4) Comments on the results\n",
    "\n",
    "The findings in our novel results align with those obtained in our initial application:\n",
    "\n",
    "* The same set of drugs were prioritised in both applications, although with minor changes in their ranking. For instance, Zoledronic acid (DB00399), previously ranked top 5, now holds the first position.\n",
    "\n",
    "* Eight out of the top 10 ranked genes from the previous results were also prioritised in the new results, with slight variations in their ranking. Notably, two novel genes, ZBTB17 and GABPA, were prioritised, ranking 3rd and 6th, respectively. Interestingly, ZBTB17 codes for a protein involved in the regulation of c-myc, which is known to play a crucial role in leukogenesis (1).  GABPA encodes a subunit of the GA-binding protein transcription factor, implicated in chronic myelogenous leukemia development (2).\n",
    "\n",
    "The incorporation of the metabolic network as a new layer in the multilayer network allowed us to prioritize metabolic reactions. The top-ranked reactions were linked to Terpenoid backbone biosynthesis, Acylglycerides metabolism, Glycerophospholipid metabolism, Cholesterol metabolism, and Inositol phosphate metabolism. Notably, Cholesterol metabolism and Inositol phosphate metabolism have been documented as being altered in leukemia (3,4). More broadly, disrupted lipid metabolism is increasingly recognized as a hallmark of cancer (5).\n",
    "\n",
    "\n",
    "    (1) Salvatori B, Iosue I, Djodji Damas N, Mangiavacchi A, Chiaretti S, Messina M, Padula F, Guarini A, Bozzoni I, Fazi F, Fatica A. Critical Role of c-Myc in Acute Myeloid Leukemia Involving Direct Regulation of miR-26a and Histone Methyltransferase EZH2. Genes Cancer. 2011 May;2(5):585-92. doi: 10.1177/1947601911416357. PMID: 21901171; PMCID: PMC3161419.\n",
    "    \n",
    "    (2) Yang, Z. F., Zhang, H., Ma, L., Peng, C., Chen, Y., Wang, J., ... & Rosmarin, A. G. (2013). GABP transcription factor is required for development of chronic myelogenous leukemia via its control of PRKD2. Proceedings of the National Academy of Sciences, 110(6), 2312-2317.\n",
    "    \n",
    "    (3) Zhao, L., Zhan, H., Jiang, X., Li, Y., & Zeng, H. (2019). The role of cholesterol metabolism in leukemia. Blood Science, 1(01), 44-49.\n",
    "    \n",
    "    (4) Bagacean, C., Iuga, C. A., Bordron, A., Tempescul, A., Pralea, I. E., Bernard, D., ... & Renaudineau, Y. (2022). Identification of altered cell signaling pathways using proteomic profiling in stable and progressive chronic lymphocytic leukemia. Journal of Leukocyte Biology, 111(2), 313-325.\n",
    "    \n",
    "    (5) Karagiota, A., Chachami, G., & Paraskeva, E. (2022). Lipid metabolism in cancer: The role of acylglycerolphosphate acyltransferases (AGPATs). Cancers, 14(1), 228."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
